src/feature/hs/hs_common.c

```diff
@@ -1723,45 +1714,38 @@ hs_get_extend_info_from_lspecs(const smartlist_t *lspecs,
     }
   } SMARTLIST_FOREACH_END(ls);
 
+  /* Choose a preferred address first, but fall back to an allowed address.
+   * Skip IPv6 if we're not direct as relays only extend via IPv4. */
+  fascist_firewall_choose_address_ls(lspecs, 0, &ap, direct_conn);
+
   /* Legacy ID is mandatory, and we require IPv4. */
-  if (!have_v4 || !have_legacy_id) {
+  if (!tor_addr_port_is_valid_ap(&ap, 0) || !have_legacy_id) {
```

Comment:
The comment in the deleted else block below now belongs here, but it should say:
 /* If we can't reach any IP address, or we are missing a legacy ID, return NULL. */